Userland/Libraries/LibJIT/ELFBuild.cpp

```diff
@@ -0,0 +1,137 @@
+}
+
+u32 StringTable::insert(StringView str) noexcept
+{
+    // The offsets for sh_name and st_name are 32-bit unsigned integers, so it
+    // won't make sense to address a string table bigger than what u32 can
+    // provide.
+    VERIFY(m_data.size() < NumericLimits<u32>::max());
+    auto const offset = static_cast<u32>(m_data.size());
+
+    auto const final_size = m_data.size() + str.length() + 1;
+    VERIFY(final_size < NumericLimits<u32>::max());
+
+    m_data.ensure_capacity(m_data.size() + str.length() + 1);
+
+    for (auto ch : str) {
+        VERIFY(ch != 0);
+        m_data.unchecked_append(ch);
+    }
+    m_data.append(0);
+    return offset;
+}
+
+FixedArray<u8> build_elf_image(u64 shstrndx, ReadonlySpan<Section> sections)
+{
+    Checked<u64> final_image_size = sizeof(Elf64_Ehdr);
+    Vector<u64> section_offsets;
+    section_offsets.ensure_capacity(sections.size());
+
+    auto const sections_begin = final_image_size.value_unchecked();
+    final_image_size += sizeof(Elf64_Shdr) * sections.size();
+
+    for (auto const& section : sections) {
+        auto const offset = final_image_size.value();
+        section_offsets.unchecked_append(offset);
+        if (section.data.has_value()) {
+            final_image_size += section.data.value().size();
+        }
+    }
+
+    auto image = MUST(FixedArray<u8>::create(final_image_size.value()));
+
+    {
+        auto section_headers = Span<Elf64_Shdr> {
+            reinterpret_cast<Elf64_Shdr*>(image.span().offset_pointer(sections_begin)),
+            sections.size(),
+        };
+        for (size_t i = 0; i < sections.size(); ++i) {
+            section_headers[i] = sections[i].header;
+            section_headers[i].sh_offset = section_offsets[i];
+            if (sections[i].data.has_value()) {
+                auto const data = sections[i].data.value();
+                auto const data_in_elf = image.span().slice(section_offsets[i], data.size());
+                data.copy_to(data_in_elf);
+                section_headers[i].sh_size = data.size();
+            }
+        }
+    }
+
+    {
+        auto* const final_elf_hdr = reinterpret_cast<Elf64_Ehdr*>(image.data());
+        final_elf_hdr->e_ident[EI_MAG0] = 0x7f;
+        final_elf_hdr->e_ident[EI_MAG1] = 'E';
+        final_elf_hdr->e_ident[EI_MAG2] = 'L';
+        final_elf_hdr->e_ident[EI_MAG3] = 'F';
+        final_elf_hdr->e_ident[EI_CLASS] = ELFCLASS64;
+        // FIXME: This is platform-dependent. Any big-endian host will write the data in
+        // MSB format, so the EI_DATA field should be set to ELFDATA2MSB.
+        final_elf_hdr->e_ident[EI_DATA] = ELFDATA2LSB;
+        final_elf_hdr->e_ident[EI_VERSION] = EV_CURRENT;
+        // XXX: This is platform-dependent. The host must set the OSABI to the
+        // same it's being executed in. I'm not sure if this matters for GDB
+        // compatibility, though.
+        final_elf_hdr->e_ident[EI_OSABI] = ELFOSABI_SYSV;
```

Comment:
Serenity images actually end up using ELFOSABI_GNU because of some quirk in some special sections we don't tell the compiler to skip. But this image is probably fine as-is :)

Reply:
Should I... add a fixme or a comment? I'm not sure how to respond to this.

Reply:
Will be leaving this as-is. I just set the abi to `ELFOSABI_AIX` and GDB still displays a correct backtrace when I make the program crash.